job-info: add JSON_DECODE & CURRENT flags, deprecate job-info.update-lookup

chu11 opened the pull request:
This PR could be split up a bit more, but for now it's one PR.

This adds several new lookup flags, and with their addition, the `job-info.update-lookup` RPC target can be deprecated.

The one "tricky" part was how to grab cached data from current `update-watch`es being done.  Perhaps the lookup of that "cached" info could just be dropped as it probably has minimal use.

I decided to not support `job-info.update-watch` within the original `job-info.lookup` RPC for several reasons.  Most notably because `lookup` can take multiple keys as input and `job-info.update-watch` takes only one key.  I could make it so a WATCH flag only works with one key.  Or we could have a "re-direct to this RPC" like we do with KVS watch.  But I didn't delve too much into this design yet.

~~I then updated PR #5592 to be based on this PR.~~  PR #5635 then follows this one up

TODO RFC spec 41 which I was going to hold off on until this gets some minimal review.

chu11 commented:
hmmm, should I have used the term "UPDATED" vs "UPDATE" as the flag?  UPDATE may imply "to update" vs "UPDATED" is "updates have been applied".

garlick commented:
Instead of update(d), maybe "current"?

garlick commented on src/modules/job-info/lookup.c:
```diff
@@ -288,17 +406,29 @@ static int check_keys_for_eventlog (struct lookup_ctx *l)
+    if (!l->allow || (l->flags & FLUX_JOB_LOOKUP_UPDATE)) {
+        size_t index;
+        json_t *key;
+        json_array_foreach(l->keys, index, key) {
```
add space before open paren

garlick commented on src/modules/job-info/lookup.c:
```diff
@@ -288,17 +406,29 @@ static int check_keys_for_eventlog (struct lookup_ctx *l)
+        size_t index;
+        json_t *key;
+        json_array_foreach(l->keys, index, key) {
+            const char *keystr;
+            if (!(keystr = json_string_value (key))) {
+                errno = EINVAL;
+                return -1;
+            }
+            if (streq (keystr, "eventlog"))
+                return 0;
         }
```
It would probably be good to check that the request `keys` is an array of strings right in the request handler rather than deeper in the code, for clarity and to avoid repeating the same check in several places.  Then this could just be:
```c
if (streq (json_string_value (key), "eventlog"))
    return 0;
```

garlick reviewed (commented):
First pass with some comments inline.

Also: we just added `job-info.update-lookup` in 0.56.  Would it be reasonable to just update users and remove or were you thinking that's another PR?

chu11 commented:
re-pushed going with JSON_DECODE and CURRENT as the flag names.

chu11 commented on src/common/libjob/job.h:
```diff
@@ -30,6 +30,14 @@ enum job_event_watch_flags {
+     * json object) instead of a string.
+     */
+    FLUX_JOB_LOOKUP_DECODE = 1,
+};
```
hmmm.  Good point, but it is useful to have around instead of hard coding.  I'm not sure the python bindings can use it if it's a non-public header?

chu11 commented on src/modules/job-info/lookup.c:
```diff
@@ -156,11 +159,105 @@ static int lookup_keys (struct lookup_ctx *l)
+    }
+
+    if (!(updated_str = json_dumps (update_object, 0)))
+        goto error;
```
considered this, it is mildly annoying to return the object given the JSON_DECODE logic that later follows in the caller.  I'll add as a fixup commit, lemme know what you think.

chu11 commented:
re-pushed with tweaks per comments above.  There's one fixup patch that adds an update that I'm a little meh on, so I left it as a fixup for the time being so it can be reviewed as a "consideration".

grondo commented on src/common/libjob/job.h:
```diff
@@ -35,6 +35,11 @@ enum job_lookup_flags {
+     * updates for those fields
+     * - currently works for jobspec and R
+     */
+    FLUX_JOB_LOOKUP_CURRENT = 2,
```
A question about this commit. The comment here indicates "works for jobspec and R" but I only see R handled specifically in `lookup_current`. Maybe a comment in `lookup_current` as to why jobspec is not handled there would be helpful.

chu11 replied:
ahh comment is probably a development history mistake, see #5635

grondo commented on src/cmd/flux-job.c:
```diff
@@ -3427,23 +3427,19 @@ int cmd_info (optparse_t *p, int argc, char **argv)
-                                 "flags", 0))
-            || flux_rpc_get_unpack (f, "{s:o}", key, &o) < 0)
+                                 "keys", key,
+                                 "flags", FLUX_JOB_LOOKUP_CURRENT))
```
Use of this flag requires a current version of flux-core. I'm not sure if it is worth it, but should this call fall back to `update-lookup` if the RPC returns `EPROTO`?

chu11 replied:
ugh, It ends up I didn't check "flags" in older versions b/c flags wasn't used.  So I guess a caller would just get the original version.

Could just keep on calling `job-info.update-lookup`?  But I'm not sure the pros of that outweigh the cons.  I figure most users, if they are using `flux job info`, base R == updated R anyways?

grondo replied:
>  I figure most users, if they are using flux job info, base R == updated R anyways?

Yes, there will be an updated R only if the expiration is adjusted, which is going to pretty rare. Also, the command would have to be run against a downrev parent instance to hit the failure, and at least this would be a failure I think instead of returning wrong information. So perhaps we just keep things as is and hope for the best? Over time it won't be a problem anymore... (However, something to keep in mind when adding future interfaces)

grondo commented on src/modules/job-info/lookup.c:
```diff
@@ -289,9 +283,25 @@ static void info_lookup_continuation (flux_future_t *fall, void *arg)
                 goto error;
-            s = current_value;
+            /* tiny optimization, to skip unnecessary json_dumps() and
+             * fallthrough to code below
```
I'd be fine dropping this optimization for now and only address performance if it becomes an issue.

grondo reviewed (commented):
I realized we should probably have merged this one before I split up `flux-job.c`, so giving this a review now (and sorry this languished so long!).

My one thought is that unfortunately the `job-info.update-lookup` RPC made it into released versions of flux-core, so we have to be careful about switching `flux job` and other users to the new flags in case a newer flux-core is used with an older. For `flux-job` however, I don't know how likely it is that there would be an issue so maybe it is fine?

grondo reviewed (approved):
Given the discussion about the `flux job info` change, this can probably go in as is when the fixup commit is dropped.

chu11 commented:
just re-pushed.  apologies, i began cleaning up more thinking that the split of `flux-job.c` had already been merged into master, whoops.  I hope I didn't clean up too much and make the eventual rebasing of PR #5747 more difficult than it needs to be.

- removed the fixup
- tweaked that one comment per @grondo comment above
- noticed some stupid typos ... https://github.com/flux-framework/flux-core/pull/5633/commits/ad606a818928c7c0a90613c1456db7b5d706095e  
- given realization above that `job-info.lookup` doesn't check for valid flags, added that check in https://github.com/flux-framework/flux-core/pull/5633/commits/cbe957de17b3503f486ed24055ed63ae30d772c3 (and added appropriate updates when new flags added)
- realized needed some extra coverage https://github.com/flux-framework/flux-core/pull/5633/commits/a43f24f97ad307a0cc9bc8d0aa24b140ab7fce2c given issues above

this PR is getting long, could split it up?

grondo commented:
Since splitting up `flux-job.c` just copy and pastes code it should be easy to fixup after all this is merged (just recopy the relevant code I hope)

If you want to split this into a separate PR that is fine with me. Thinking about each PR as a "release note entry" though, what would that be?

chu11 commented:
> If you want to split this into a separate PR that is fine with me. Thinking about each PR as a "release note entry" though, what would that be?

It'd just be a "cleanup" PR, so I guess nothing useful.  Only offering to split up if it would make reviewing easier ... but I now see that you approved the PR pre-my recent changes.  So I guess that doesn't matter as much now :-)
